### PR TITLE
Innocuous housekeeping changes in the partially determined haplotypes code

### PR DESCRIPTION
**Pull request by davidbenjamin**
@jamesemery Getting into the `PartiallyDeterminedHaplotypeComputationEngine` now.  No change in output here.  I don't think there's anything controversial here.  The commit history is discrete and pristine.

My next DRAGEN PR might be where the fun begins.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 1]
```diff
@@ -937,29 +862,87 @@ public List<List<Tuple<Event,Boolean>>> getVariantGroupsForEvent(final List<Tupl
...
 
         public boolean causesBranching() {
-            return variantsInBitmapOrder.size() > 1;
+            return eventsInBitmapOrder.size() > 1;
         }
 
         //Print The event group in Illumina indexed ordering:
         public String toDisplayString(int startPos) {
-            return "EventGroup: " + variantsInBitmapOrder.stream().map(vc -> PartiallyDeterminedHaplotype.getDRAGENDebugEventString(startPos).apply(vc)).collect(Collectors.joining("->"));
+            return "EventGroup: " + dragenString(eventsInBitmapOrder, startPos);
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:870]: slightly more informative name than "dragen string"

**Review by jamesemery — COMMENTED** [review]
Good refactor overall... almost all minor comments with one request for a unit test for a method that should have had one to begin with (hopefully not to onerous for you)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 2]
```diff
@@ -818,40 +745,40 @@ public EventGroup(final Event ... events) {
          * @param disallowedEvents Pairs of events disallowed
          * @return false if the event group is too large to process
          */
-        public boolean populateBitset(List<List<Event>> disallowedEvents) {
-            if (variantsInBitmapOrder.size() > MAX_VAR_IN_EVENT_GROUP) {
-                return false;
-            }
-            if (variantsInBitmapOrder.size() < 2) {
-                return true;
+        public void populateBitset(List<List<Event>> disallowedEvents) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:748]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 3]
```diff
@@ -818,40 +745,40 @@ public EventGroup(final Event ... events) {
...
+            if (eventsInBitmapOrder.size() > MAX_VAR_IN_EVENT_GROUP || eventsInBitmapOrder.size() < 2) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:749]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:870]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 4]
```diff
@@ -186,63 +132,53 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
...
          * Overall Loop:
          * Iterate over every cluster of variants with the same start position.
          */
-        for (int indexOfDeterminedInR = 0; indexOfDeterminedInR < entriesRInOrder.size(); indexOfDeterminedInR++) {
-            Map.Entry<Integer, List<Event>> variantSiteGroup = entriesRInOrder.get(indexOfDeterminedInR);
-            if (debugSite) System.out.println("working with variants of the group: " + variantSiteGroup);
-            // Skip
-            if (entriesRInOrder.get(indexOfDeterminedInR).getKey() < callingSpan.getStart() || entriesRInOrder.get(indexOfDeterminedInR).getKey() > callingSpan.getEnd()) {
-                if (debugSite) System.out.println("Skipping determined hap construction! otside of span: "+callingSpan);
+        for (final int start : variantsByStartPos.keySet()) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:154]: added that comment

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 5]
```diff
@@ -937,29 +862,87 @@ public List<List<Tuple<Event,Boolean>>> getVariantGroupsForEvent(final List<Tupl
...
+        return dragenString(events, refStart,"->");
+    }
+
+    private static String dragenString(final Collection<Event> events, final int refStart, final CharSequence delimiter) {
+        return events.stream()
+                .map(PartiallyDeterminedHaplotype.getDRAGENDebugEventString(refStart))
+                .collect(Collectors.joining(delimiter));
+    }
+
+    private static double dragenStart(final Event event) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:903]: done and done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 6]
```diff
@@ -937,29 +862,87 @@ public List<List<Tuple<Event,Boolean>>> getVariantGroupsForEvent(final List<Tupl
...
-            variantsInBitmapOrder.addAll(other.variantsInBitmapOrder);
-            variantContextSet.addAll(other.variantsInBitmapOrder);
+            eventsInBitmapOrder.addAll(other.eventsInBitmapOrder);
+            eventSet.addAll(other.eventsInBitmapOrder);
             allowedEvents = null;
             return this;
         }
     }
+
+    private static String dragenString(final Collection<Event> events, final int refStart) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:893]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java [hunk 7]
```diff
@@ -939,4 +939,9 @@ public AssemblyResultSet generateEmptyLLocalAssemblyResult(final AssemblyRegion
 
         return resultSet;
     }
+
+    private static String dotFileName(final Haplotype refHaplotype, final int kmerSize, final String suffix) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java:943]: I wrote said tool.  Was it overkill?

File: src/main/java/org/broadinstitute/hellbender/utils/Utils.java [hunk 8]
```diff
@@ -824,6 +824,12 @@ public static void validate(final boolean condition, final Supplier<String> msg)
         }
     }
 
+    public static void printIf(final boolean condition, final Supplier<String> msg){
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/Utils.java:827]: I will leave it for future idealistic times.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 9]
```diff
@@ -355,82 +284,91 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
...
+        if (!pileupArgs.determinePDHaps) {
             // Setting a boolean on the source-set to indicate to downstream code that we have PD haplotypes
             sourceSet.setPartiallyDeterminedMode();
         }
-        if (debugSite ) System.out.println("Returning "+outputHaplotypes.size()+" to the HMM");
+        Utils.printIf(debug, () -> "Returning "+outputHaplotypes.size()+" to the HMM");
         return sourceSet;
     }
 
+    private static List<Event> makeFinalListOfEventsInOrder(final AssemblyResultSet sourceSet, final Set<Event> badPileupEvents,
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:320]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 10]
```diff
@@ -68,114 +70,58 @@ public class PartiallyDeterminedHaplotypeComputationEngine {
...
         // Iterate over all events starting with all indels
 
-        List<List<Event>> disallowedPairs = smithWatermanRealignPairsOfVariantsForEquivalentEvents(referenceHaplotype, aligner, swParameters, debugSite, eventsInOrder, vcsAsList);
-        if (debugSite) {
-            System.out.println("disallowed Variant pairs:");
-            disallowedPairs.stream().map(l -> l.stream().map(PartiallyDeterminedHaplotype.getDRAGENDebugEventString((int) referenceHaplotype.getStartPosition())).collect(Collectors.joining("->"))).forEach(System.out::println);
-        }
+        List<List<Event>> disallowedPairs = smithWatermanRealignPairsOfVariantsForEquivalentEvents(referenceHaplotype, aligner, args.getHaplotypeToReferenceSWParameters(), debug, eventsInOrder);
+        dragenDisallowedGroupsMessage(referenceHaplotype.getStart(), debug, disallowedPairs);
+        Utils.printIf(debug, () -> "Event groups before merging:\n"+eventGroups.stream().map(eg -> eg.toDisplayString(referenceHaplotype.getStart())).collect(Collectors.joining("\n")));
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:116]: Okay, no more change for now in that case.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 11]
```diff
@@ -302,22 +230,20 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
                     List<Event> newBranch = new ArrayList<>();
 
                     // Handle the simple case of making PD haplotypes
-                    if (!makeDeterminedHapsInstead) {
-                        for (int secondRIndex = 0; secondRIndex < entriesRInOrder.size(); secondRIndex++) {
-                            if (secondRIndex != indexOfDeterminedInR) {
+                    if (!pileupArgs.determinePDHaps) {
+                        for (final int secondStart : variantsByStartPos.keySet()) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:234]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 12]
```diff
@@ -256,42 +192,34 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
...
-                            for (int i = 1; i < groupVCs.size(); i++) { //NOTE: iterate starting at 1 here because we special case that branch at the end
-                                Set<Event> newSet = new HashSet<>(excludedVars);
-                                groupVCs.get(i).stream().filter(t -> !t.b).forEach(t -> newSet.add(t.a));
-                                newBranchesToAdd.add(newSet);
-                            }
-                            // Be careful since this event group might have returned nothing
-                            if (!groupVCs.isEmpty()) {
-                                groupVCs.get(0).stream().filter(t -> !t.b).forEach(t -> excludedVars.add(t.a));
-                            }
+                    if (!group.causesBranching()) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:195]: Okay, I reverted it.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 13]
```diff
@@ -326,24 +252,27 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
                         List<List<Event>> variantGroupsCombinatorialExpansion = new ArrayList<>();
                         variantGroupsCombinatorialExpansion.add(new ArrayList<>());
                         // We can drastically cut down on combinatorial expansion here by saying each allele is the FIRST variant in the list, thus preventing double counting.
-                        for (int secondRIndex = indexOfDeterminedInR; secondRIndex < entriesRInOrder.size(); secondRIndex++) {
-                            if (variantGroupsCombinatorialExpansion.size() > MAX_BRANCH_PD_HAPS) {
-                                if(debugSite ) System.out.println("Too many branch haplotypes ["+variantGroupsCombinatorialExpansion.size()+"] generated from site, falling back on assebmly variants!");
+                        for (final int secondStart : variantsByStartPos.keySet()) {
+                            // We reduce combinatorial expansion by saying each allele is the first variant in the list, thus preventing double counting.
+                            if (secondStart < start) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:257]: I took out the continue and tightened the loop.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 14]
```diff
@@ -480,26 +418,23 @@ private static List<List<Event>> smithWatermanRealignPairsOfVariantsForEquivalen
...
      * indels don't overlap on their anchor bases and insertions don't overlap anything except deletions spanning them or other insertions
      * at the same base.
      *
-     * @param snpsOverlap if false, don't ever evaluate snps as overlapping other snps (we do this because sometimes we need to construct artifical haps where we don't allow overlapping)
      */
-    static boolean eventsOverlapForPDHapsCode(Event e1, Event e2, boolean snpsOverlap){
-        if (!snpsOverlap && e2.isSNP() && e1.isSNP()) {
-            return false;
-        }
+    static boolean eventsOverlapForPDHapsCode(Event e1, Event e2){
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:424]: done

**Comment by davidbenjamin** [comment]
Okay @jamesemery, ready for the next round of review.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngineUnitTest.java [hunk 15]
```diff
@@ -214,4 +215,42 @@ public void testDeletionUnderlapingDeterminedBases() {
         Assert.assertEquals(result.getCigar(), TextCigarCodec.decode("10M"));
         Assert.assertEquals(result.getDeterminedPosition(), DEL_AA_105.getStart());
     }
+
+    @Test
+    public void testEventsOverlapForPDHapsCode() {
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngineUnitTest.java:219]: 👍 i think we are going to appreciate having this unit test in some vague undetermined point in the future

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java [hunk 16]
```diff
@@ -940,8 +907,48 @@ public AssemblyResultSet generateEmptyLLocalAssemblyResult(final AssemblyRegion
...
+     * . . .
+     * chr20:20-100-sequenceGraph.13.1.0.non_ref_removed.dot,
+     * where 13 is the kmer size, and the following two numbers indicate broad categories (0 for the initial read threading
+     * de Bruijn graph, 1 for the later seq graph) and individual tags, respectively
+     * . . .
+     *
+     * This class automatically increments the section whenever a new graph is printed i.e. it switches from 0 to 1 once
+     * it is asked to print a transformation of the seq graph instead of the read threading graph
+     */
+    private class DotFilePrinter {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java:909]: This looks good. Can i ask you for one (very quick) test of this? can you run this over a small section of our test files with dbug graph printing enabled just to make sure that its correctly printing the dotfiles named correctly for each assembly region and graph within the assembly window (so regionsize.kmersize for both the 10 and 25mer graphs at most sites).

Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java:909]: or if you've already done a test like that (just make 100% sure its printing correctly across a whole region) then 👍

**Review by jamesemery — APPROVED** [review]
One very light request that you can check on your own. Otherwise these responses look good to me. 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java (hunk 16 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java:909]: I checked and it's working as before.

**Pull request merged**